File: Submissions/515_Find_Largest_Value_in_Each_Tree_Row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        vector<int> result;
        if(root == nullptr) return result;
        
        queue<TreeNode*> q;
        TreeNode* node = root;
        q.push(node);
        while(!q.empty()){
            int n = q.size();
            int max_val = q.front()->val;
            for(int i = 0; i < n; ++i){
                node = q.front();
                max_val = std::max(max_val, node->val);
                if(node->left != nullptr) q.push(node->left);
                if(node->right != nullptr) q.push(node->right);
                q.pop();
            }
            result.push_back(max_val);
        }
        
        return result;
    }

};

File: Kotlin/515_Find_Largest_Value_in_Each_Tree_Row.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun largestValues(root: TreeNode?): List<Int> {
        val result: MutableList<Int> = mutableListOf()
        if(root == null) return result
        
        val q = LinkedList<TreeNode>()
        q.offer(root)
        
        while(!q.isEmpty()){
            val n = q.size
            var max_value = q.peek().`val`
            for(i in 0 until n){
                val current = q.pop()
                max_value = Math.max(max_value, current.`val`)
                if(current.left != null) q.offer(current.left)
                if(current.right != null) q.offer(current.right)
            }
            result.add(max_value)
        }
        
        return result
    }
}